src/ok.rs

```diff
@@ -0,0 +1,125 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::*;
+use std::arch::x86_64;
+use std::mem::transmute;
+use std::str::from_utf8;
+
+#[derive(StructOpt)]
+pub enum SevGeneration {
+    #[structopt(about = "AMD SEV Extensions")]
+    Sev,
+
+    #[structopt(about = "AMD SEV-ES Extensions")]
+    Es,
+}
+
+struct CpuId {
+    pub name: &'static str,
+    pub leaf: u32,
+    pub func: fn(x86_64::CpuidResult) -> (bool, Option<String>),
+}
+
+struct TestResult {
+    pub success: bool,
+    pub msg: String,
+}
+
+const CORE_CPUID: &[CpuId] = &[CpuId {
+    name: "CPU Manufacturer",
+    leaf: 0x00000000,
+    func: |res| {
+        let name: [u8; 12] = unsafe { transmute([res.ebx, res.edx, res.ecx]) };
+        let name = from_utf8(&name[..]).unwrap();
+
+        (name == "AuthenticAMD", None)
+    },
+}];
+
+const AMD_CPU_DEPENDENT_CPUIDS: &[CpuId] = &[
+    CpuId {
+        name: "Microcode support",
+        leaf: 0x80000002,
+        func: |_res| {
+            let cpu_name = {
+                let mut bytestr = Vec::with_capacity(48);
+                for cpuid in 0x8000_0002_u32..=0x8000_0004_u32 {
+                    let cpuid = unsafe { x86_64::__cpuid(cpuid) };
+                    let mut bytes: Vec<u8> = [cpuid.eax, cpuid.ebx, cpuid.ecx, cpuid.edx]
+                        .iter()
+                        .flat_map(|r| r.to_le_bytes().to_vec())
+                        .collect();
+                    bytestr.append(&mut bytes);
+                }
+                String::from_utf8(bytestr).unwrap().trim().to_string()
```

Comment:
No unwraps. You can just use a value that'll fail the check. If that's the case, the message should be updated accordingly so we know what is read.

```suggestion
                String::from_utf8(bytestr).unwrap_or("???").trim().to_string()
```

Reply:
This should also be fixed in this commit, not a later commit.